lib/Enforcers/RevngEnforcers.cpp

```diff
@@ -0,0 +1,102 @@
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include "llvm/ADT/StringMap.h"
+
+#include "revng/AutoEnforcer/AutoEnforcerLibraryRegistry.h"
+#include "revng/AutoEnforcer/AutoEnforcerTarget.h"
+#include "revng/AutoEnforcer/InspectableLLVMContainer.h"
+#include "revng/Enforcers/CompileModule.h"
+#include "revng/Enforcers/Isolate.h"
+#include "revng/Enforcers/Lift.h"
+#include "revng/Enforcers/LinkForTranslation.h"
+#include "revng/Enforcers/LinkSupport.h"
+#include "revng/Enforcers/RevngEnforcers.h"
+#include "revng/Enforcers/RootKind.h"
+#include "revng/Enforcers/TemporaryFile.h"
+
+using namespace std;
+using namespace AutoEnforcer;
+
+Granularity AutoEnforcer::RootGranularity("Root Granularity");
+Granularity
+  AutoEnforcer::FunctionsGranularity("Function Granularity", RootGranularity);
+
+Kind AutoEnforcer::Binary("Binary", &RootGranularity);
+
+Kind AutoEnforcer::Object("Object", &RootGranularity);
+Kind AutoEnforcer::Translated("Translated", &RootGranularity);
+Kind AutoEnforcer::Dead("Dead", &RootGranularity);
+
+Kind AutoEnforcer::CFepper("CFepper", &FunctionsGranularity);
+Kind AutoEnforcer::ABIEnforced("ABIEnforced", Isolated, &FunctionsGranularity);
+
+llvm::Error RevngModelAutoEnforcerWrapper::save(llvm::StringRef Path) const {
+  return serializeToFile(Model.getReadOnlyModel(), Path);
+}
+
+llvm::Error RevngModelAutoEnforcerWrapper::load(llvm::StringRef Path) {
+  if (not llvm::sys::fs::exists(Path))
+    return llvm::Error::success();
+
+  auto MaybeModel = TupleTree<model::Binary>::loadFromFile(Path);
+  if (not MaybeModel)
+    return llvm::make_error<llvm::StringError>("Could not parse model",
+                                               MaybeModel.getError());
+
+  Model = std::move(*MaybeModel);
+  return llvm::Error::success();
+}
+namespace AutoEnforcer {
+llvm::Error autoEnforcerRevngConfigurationCallback(const PipelineLoader &Loader,
+                                                   LLVMEnforcer &NewPass) {
+  using Wrapper = RevngModelAutoEnforcerWrapper;
+  auto MaybeModelWrapper = Loader.getContext().get<Wrapper>(Wrapper::Name);
+  if (not MaybeModelWrapper)
+    return MaybeModelWrapper.takeError();
+
+  auto &ModelWrapper = (*MaybeModelWrapper)->getModelWrapper();
+  NewPass.emplacePass<LoadModelEnforcerPass>(ModelWrapper);
+  return llvm::Error::success();
+}
+
+const char RevngModelAutoEnforcerWrapper::ID = '0';
+
+class RevngRegistryLibrary : public AutoEnforcerLibraryRegistry {
```

Comment:
You're creating a class with methods that populate various registries.
An alternative would be to have a more *declarative* approach and just have a global variable of type `AutoEnforcerLibraryRegistry` whose constructor takes all the necessary arguments.

Do you see a use case in which we want to conditionally register stuff or make any other sort of computation?

Or to be even more aggressive, what do you think of letting individual containers, enforcers and kinds to self-register? Is the registration order relevant here?

I'm trying to see how much we can dematerialize this class.

Reply:
the issue is that for complex enforcers and containers it may be needed to look into the global variables, and thus i opted to perform all those things in the same place, so that at least one could have a insight of the structure of library by simply looking at that registry. I think i can change it so that the average case can be declared on it's own, but i don't think there will be so many enforcers that we don't wont a centralized location for them

Reply:
> the issue is that for complex enforcers and containers it may be needed to look into the global variables

What do you mean by that? If named globals are global variables too you can reference them from enforcers.

>  i don't think there will be so many enforcers that we don't wont a centralized location for them

Yeah, maybe. But why centralize if not strictly necessary?
I'd like to follow the LLVM approach for passes as much as possible. This also enables you to create an enforcer with no public headers at all and keep the friction for creating a new enforcer low.

Reply:
named globals are not globals in the cpp sense of term but globals in the sense that can be used globally from anywhere in the autoenforcer. The requirement is just that their lifetime is superior to the autoenforcer

ok, i will decentralize them, may take a bit this change toh.